### PR TITLE
Rover Sailboat: refactor the mainsail/wingmast/mast_rotation logic

### PR DESCRIPTION
**Pull request by kwikius**
Rover Sailboat: refactor the mainsail/wingmast/mast_rotation logic and make non const sailboat functions private to the sailboat class. Refactor saves around 128 bytes of text image size (in SITL anyway) and easier to read the source code

### CODE REVIEW

File: Rover/sailboat.cpp [hunk 1]
```diff
@@ -172,57 +172,24 @@ void Sailboat::init_rc_in()
...
+void Sailboat::set_pilot_desired_mainsail()
 {
     // no RC input means mainsail is moved to trim
     if ((rover.failsafe.bits & FAILSAFE_EVENT_THROTTLE) || (channel_mainsail == nullptr)) {
-        mainsail_out = 100.0f;
-        wingsail_out = 0.0f;
-        mast_rotation_out = 0.0f;
-        return;
+       relax_sails();
+    }else{
```
Comment (IamPete1) [inline Rover/sailboat.cpp:182]:
```suggestion
    } else {
```

File: Rover/sailboat.cpp [hunk 2]
```diff
@@ -232,55 +199,41 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
     const float wind_dir_apparent_sign = is_negative(wind_dir_apparent) ? -1.0f : 1.0f;
 
     //
-    // mainsail control
+    // mainsail control.
     //
-
+    // mainsail_out represents a range from 0 to 100
+    float mainsail_out = 100.f;
```
Comment (IamPete1) [inline Rover/sailboat.cpp:205]: I fund the `.` harder to read than `.0`, we don't really need the `f` either. (I know there all over already)

File: Rover/sailboat.cpp [hunk 3]
```diff
@@ -310,7 +263,45 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
...
+                                                                        rover.g.speed_cruise,
+                                                                        rover.g.throttle_cruise * 0.01f,
+                                                                        rover.G_Dt);
+    } else {
+        throttle_out = 0.0f;
+    }
+
+    if (motor_state == UseMotor::USE_MOTOR_ALWAYS) {
+       relax_sails();
+    }else{
```
Comment (IamPete1) [inline Rover/sailboat.cpp:302]:
```suggestion
    } else {
```

**Review by IamPete1 — COMMENTED** [review]
Nice improvement! Looks like there should not be any change in behavior?

**Comment by kwikius** [comment]
> Nice improvement! Looks like there should not be any change in behavior?

There is a change in behaviour from https://github.com/ArduPilot/ardupilot/blob/master/Rover/mode_manual.cpp#L38 , which I think was a bug? So I changed behaviour to as if `g2.motors.set_mast_rotation(desired_mast_rotation);

**Comment by kwikius** [comment]
Not sure if I messed up the PR. Maybe shouldn't have incorporated latest changes from master?

**Comment by kwikius** [comment]
Reset to head at original PR prior to rebase

File: Rover/sailboat.h [hunk 4]
```diff
@@ -18,6 +18,15 @@
...
 class Sailboat
 {
+
+   friend class Rover;
+   friend class Mode;
+   friend class ModeManual;
+   friend class ModeHold;
+   friend class ModeLoiter;
+   friend class ModeAcro;
+   friend class RC_Channel_Rover;
```
Comment (IamPete1) [inline Rover/sailboat.h:28]: I don't think these are needed for this PR? There are no changes to most of those classes.

**Review by IamPete1 — COMMENTED** [review]
Looking good, just that friend stuff and I think we can merge.

File: Rover/sailboat.h (hunk 4 above)
Comment (kwikius) [reply Rover/sailboat.h:28]: I modified Sailboat class member function access to make all the non const Sailboat member functions private and so those classes calling private Sailboat functions ( ie Sailboat state modifying functions)  require friend access to Sailboat. The overall idea is limit who can change the state as much as possible. If functions are public, anyone can change state thus making debugging harder.  The precedent is similar use of friends in Rover class https://github.com/ArduPilot/ardupilot/blob/master/Rover/Rover.h#L77 

Shall I revert it to those non const member functions of Sailboat being public?

Comment (IamPete1) [reply Rover/sailboat.h:28]: Ah, I see. I had not appreciated the change in scope of those functions. I think my preference would be to keep the current structure. 

They key difference is that all of the params and internal state variables are currently private. So cannot be changed externally. It we add friends then they can all reach in any change things they probably shouldn't. EG you should call `set_motor_state` not just change `motor_state`. That is currently enforced.

If there are things that should defiantly not be able to change anything (those class which you did not have to add as friends) then we could go via a const reference to the class. But i'm not really sure there are any that would need this?

**Review by IamPete1 — APPROVED** [review]
Looks good, Thanks, I have just added `Rover: Sailboat` to start of the last commit message.

**Review by rmackay9 — APPROVED** [review]

**Pull request merged**